### ISSUE TITLE
Build Scraper For SC Public Index

### ISSUE THREAD

**Issue by benjibuiltit** [Police-Data-Accessibility-Project/planning#10]
All counties in SC appear to use the same platform. You can get to each counties records from this page... https://www.sccourts.org/caseSearch/

I'd like to see a scraper built for this, as the single scraper could cover the entire state.

**Comment by benjibuiltit** [comment]
Happy to take this on as I'm already working on it. If someone wants to assign it to me. I really just created this issue as I wanted to track it to ensure we're not duplicating work.

**Comment by benjibuiltit** [comment]
TLDR; I believe SC uses a proprietary Public Index portal that employs active measures to detect and redirect bot traffic including scrapers.

I've reopened this issue to document some of my findings in the limited time I've spent working towards this.

Navigating the SC Public Index begins [here](https://www.sccourts.org/caseSearch/). This page links out to the Public Index system for every county in South Carolina. After selecting a county you'll be taken to the Public Index portal, and the initial page will be a disclaimer. I didn't take the time to actually read this wall of text until a few hours after working on a scraper, but included in the disclaimer is this statement about data scraping...

> Access to the South Carolina Judicial Department Public Index web sites by a site data scraper or any similar software intended to discover and extract data from a website through automated, repetitive querying for the purpose of collecting such data is expressly prohibited.

I'm skeptical at how enforceable this is considering it's all public data. More-so with the assumption that someone scraping has no ill intent, and is not overloading any servers making the site inaccessible to non-automated users. Alas, I'm no lawyer so... 🤷

The portal itself begins on the disclaimer/individual county pages and can be identified by this footer at the bottom of the page. 

> CMSWeb 6.1 © 2019 South Carolina Judicial Branch • All rights reserved

After briefly searching for CMSWeb 6.1 online all results seem related to this public index portal and limited to South Carolina, leading me to believe it's bespoke software developed and maintained by the state.

Lastly, when working to develop a scraper using Puppeteer, including using a stealth plugin for Puppeteer, the portal is still detecting automation via some means and refuses to present even the disclaimer page. The automation script is redirected back to the main county selection page anytime you attempt to navigate either directly using the Puppeteer API or indirectly (through clicking a link). In addition to using the stealth plugin for Puppeteer, I made sure JS was enabled and the same cookies were being set in the headless browser as are in a live session. I also observed bot related request headers when analyzing network requests the app makes. I'll update this with more details regarding those headers when I can.

**Comment by pelzhaus** [comment]
@benjibuiltit yeah this does look quite gnarly! Having scraped similar sites to this before, my guess is in making requests you'll have to get the right combination of cookies & headers to step through the pages. I'm seeing four layers:
1. GET to https://www.sccourts.org/caseSearch/
2. GET to https://publicindex.sccourts.org/<county>/publicindex/ (which is the validation page)
3. POST to https://publicindex.sccourts.org/<county>/publicindex/ which 302s to /<county>/PublicIndex/PISearch.aspx
4. POST for a particular search, which has dozens of data fields

Probably not impossible to crack but certainly time consuming!

**Comment by benjibuiltit** [comment]
Indeed. I think it's just a tedious process of matching everything up. I've been busy and haven't had a chance to get back to it, but I'm hoping to soon. Someone in the slack group called out attempting to set the referrer header so that's on my to-try list as well.

**Comment by sarachour** [comment]
Thought i should mention -- I was able to get to the disclaimer page for Williamsburg from ('https://www.sccourts.org/caseSearch/') by injecting javascript through a chrome extension. One possible approach here would be to write the scraper as a background script that automatically steers the browser to collect data when triggered. I can push a bit further if you all are interested.

**Comment by benjibuiltit** [comment]
Hi Sara, I had considered a similar approach as well to just execute a script from within your browser. As far as I understand, this is more or less what you're doing via a Chrome Extension. Feel free to correct me if I'm wrong / clarify your efforts.

I think this would definitely circumvent any type of bot detection, but I'm not sure how valuable it would be long term if it cannot be automated (ran on cron, etc) and requires manual initiation to harvest the data. With that said, anything is better than nothing, and I don't want to discourage you from it as I'm eager to see the data myself.

**Comment by rbarosky** [comment]
@benjibuiltit 

I did some checking on the South Carolina statewide court records portal at https://www.sccourts.org/caseSearch/.

Here are a few things I've noticed:

1. When you go to one of the counties and get to the accept/decline disclaimer page (just using desktop Chrome browser on Windows 10), sometimes you get an HTML page response and sometimes you get an obfuscated JS response that then loads the HTML into the page. The delay is noticable when loading the page, and you can notice which approach is taken.

2. Capturing the traffic in Chrome's network tab, I copied full curl requests from when the full HTML page was loaded to see if I could use that same curl request on another server, and get the HTML response. Usually, it resulted in the JS response, but sometimes the full HTML directly. However, in those cases where the full HTML was returned directly, subsequent requests using the same curl response, never resulted in the direct HTML again. It seems there may not be anything tied to requiring the same IP between requests, but from this it's likely that new cookies may be set and as well, an actual browser is in fact needed to execute JS which may modified cookies or other info for subsequent requests.

So, just for some added context.

3. I wanted to get an idea of whether repeated requests may be blocked. I went to the Abbeville search page in my browser, and used a Chrome plugin to refresh the page after every 5 seconds, and left it run for several hours. After I came back, the site had not in fact blocked me. It's possible only certain types of requests may trigger blocking, like actual search requests or case details pages. 


My suggestion: You mentioned you used Puppeteer in trying to do some scraping setup and weren't able to get the disclaimer page to load.

Can you try adding some kind of delay before checking/extracting/using the browser's response? That way, it might be the case, that you can wait for the browser to do some work as the obfuscated JS executes and the full page is loaded. I think that's worth trying.

Let me know if that works, if you have some other idea, if you tried that already, or saw something that made you think that wouldn't be useful.

**Comment by rbarosky** [comment]
@benjibuiltit ?
